src/workerd/util/url-wrapper.c++

```diff
@@ -0,0 +1,139 @@
+#include <workerd/util/url-wrapper.h>
+#include <workerd/jsg/jsg.h>
+
+namespace workerd {
+
+namespace {
+kj::OneOf<kj::String, kj::Url, jsg::Url> initialize(UrlWrapper::Mode mode, kj::StringPtr input) {
+  switch (mode) {
+    case UrlWrapper::Mode::LEGACY: {
+      return kj::str(input);
+    }
+    case UrlWrapper::Mode::STANDARD: {
+      return JSG_REQUIRE_NONNULL(jsg::Url::tryParse(input), TypeError, "Invalid URL");
+    }
+  }
+  KJ_UNREACHABLE;
+}
+}  // namespace
+
+UrlWrapper UrlWrapper::kLegacyFake = UrlWrapper::legacy("https://fake-host/");
+UrlWrapper UrlWrapper::kStandardFake = UrlWrapper::standard("https://fake-host/");
+
+UrlWrapper::UrlWrapper(Mode mode, kj::StringPtr input)
+   : inner(initialize(mode, input)) {}
+
+UrlWrapper::UrlWrapper(kj::OneOf<kj::Url, jsg::Url> inner)
+    : inner(kj::mv(inner)) {}
+
+void UrlWrapper::ensureParsed() {
+  KJ_SWITCH_ONEOF(inner) {
+    KJ_CASE_ONEOF(url, jsg::Url) {
+      // Parsed!
+    }
+    KJ_CASE_ONEOF(url, kj::Url) {
+      // Parsed!
+    }
+    KJ_CASE_ONEOF(str, kj::String) {
+      inner = JSG_REQUIRE_NONNULL(kj::Url::tryParse(str), TypeError, "Invalid URL");
```

Comment:
Aren't you going to need to pass some options to `tryParse()`? All of the uses in `workerd/api/http.c++` pass options.

Reply:
Probably. I was actually hoping to see if I could avoid it :-) Once I start plugging in the actual use I was planning to tweak this as necessary.